src/optimizer/join_elimination.cpp

```diff
@@ -0,0 +1,206 @@
+#include "duckdb/optimizer/join_elimination.hpp"
+
+#include "duckdb/catalog/catalog_entry/table_catalog_entry.hpp"
+#include "duckdb/function/function_binder.hpp"
+#include "duckdb/parser/parsed_data/vacuum_info.hpp"
+#include "duckdb/planner/binder.hpp"
+#include "duckdb/planner/column_binding_map.hpp"
+#include "duckdb/planner/expression/bound_aggregate_expression.hpp"
+#include "duckdb/planner/expression/bound_columnref_expression.hpp"
+#include "duckdb/planner/expression/bound_constant_expression.hpp"
+#include "duckdb/planner/expression_iterator.hpp"
+#include "duckdb/planner/operator/logical_comparison_join.hpp"
+#include "duckdb/planner/operator/logical_filter.hpp"
+#include "duckdb/planner/operator/logical_get.hpp"
+#include "duckdb/planner/operator/logical_order.hpp"
+#include "duckdb/planner/operator/logical_simple.hpp"
+#include "duckdb/storage/table_storage_info.hpp"
+
+namespace duckdb {
+unique_ptr<LogicalOperator> JoinElimination::Optimize(unique_ptr<LogicalOperator> op) {
+	if (!finish_collection) {
+		// We need to collect the unique constraints set at first, and we only need to collect once.
+		CollectUniqueConstraintsSet(*op);
+		finish_collection = true;
+		if (unique_constraints_set.empty()) {
+			// There are no unique constraints set, so we can return directly.
+			return op;
+		}
+	}
```

Comment:
In other optimization rules, it has similar logical. Maybe we can separate this logical into a independent rule. But I think we can optimize this later.